src/helpers/buffers/ordering_mpsc.rs

```diff
@@ -0,0 +1,418 @@
+    /// Inserts a new element to the specified position.
+    ///
+    /// When inserting, `index` needs to be in range.  Values that are in range are within `capacity`
+    /// (as provided to [`new`]) of the last value that was taken with [`take`].
+    ///
+    /// ## Panics
+    /// Panics if `index` is out of bounds or if something was previously inserted at `index`.
+    /// Panics only occur in debug builds; otherwise, a bad index will overwrite that location;
+    /// expect bad things to happen in that case.
+    /// In all builds, this panics if `msg` fails to serialize properly, which shouldn't happen.
+    ///
+    /// [`new`]: Self::new
+    /// [`take`]: Self::take
+    fn insert(&mut self, index: usize, msg: M) {
+        if cfg!(debug_assertions) {
+            let end = self.end.get();
+            assert!(
+                ((end - self.capacity.get())..end).contains(&index),
+                "Out of range at index {index} on channel \"{}\" (allowed={:?})",
+                self.name,
+                (end - self.capacity.get())..end,
+            );
+        }
+        // Translate from an absolute index into a relative one.
+        let i = index % self.capacity.get();
+        let start = i * M::SIZE_IN_BYTES;
+        let offset = start..start + M::SIZE_IN_BYTES;
+
+        let overwritten = self.added.replace(i, true);
+        debug_assert!(
+            !overwritten,
+            "Duplicate send for index {index} on channel \"{}\"",
+            self.name,
+        );
+        msg.serialize(&mut self.buf[offset]).unwrap();
+    }
+
+    /// Takes a block of elements from the beginning of the vector, or `None` if
+    /// fewer than `min_count` elements have been inserted at the start of the buffer.
+    fn take(&mut self, min_count: usize) -> Option<Vec<u8>> {
+        // Find the relative index we're starting at.
+        let i = self.end.get() % self.capacity.get();
+
+        // Find how many elements we can return, at the tail
+        let tail = self.added[i..].leading_ones();
+        // ... and if we need to wrap to the start of the buffer.
+        let wrap = if tail + i == self.capacity.get() {
+            self.added[..i].leading_ones()
+        } else {
+            0
+        };
+
+        if tail + wrap < min_count {
+            return None;
+        }
+
+        // Move `self.end` marker, clear the values in `self.added`, and
+        // return a copy of that part of `self.data` that matters.
+        self.end.incr(tail + wrap);
+        self.added[i..(i + tail)].fill(false);
+        if wrap > 0 {
+            self.added[..wrap].fill(false);
+            let mut buf = Vec::with_capacity((tail + wrap) * M::SIZE_IN_BYTES);
+            buf.extend_from_slice(&self.buf[(i * M::SIZE_IN_BYTES)..]);
+            buf.extend_from_slice(&self.buf[..(wrap * M::SIZE_IN_BYTES)]);
+            Some(buf)
+        } else {
+            Some(self.buf[(i * M::SIZE_IN_BYTES)..((i + tail) * M::SIZE_IN_BYTES)].to_owned())
```

Comment:
not for this change, but this looks like a good use case for `Bytes` (shared pointer to a slice of [u8])`?

Reply:
`Bytes` is probably a little heavyweight for this application.  We don't need the refcounting and all that stuff in `Bytes`.

I almost had this return a slice, except for the wrapping part.  It might be possible to change the calling side to make that work, but that can come later, as you say.

Reply:
oh nice, if we can eventually get to returning a slice, that's even better